lib/pbench/server/api/resources/query_apis/search_api.py

```diff
@@ -0,0 +1,221 @@
+    CONTEXT,
+    ElasticBase,
+    Schema,
+    Parameter,
+    ParamType,
+    PostprocessError,
+)
+
+
+class SearchAPI(ElasticBase):
+    """
+    Create a search request based on given query parameter.
+    """
+
+    def __init__(self, config: PbenchServerConfig, logger: Logger):
+        super().__init__(
+            config,
+            logger,
+            Schema(
+                Parameter("user", ParamType.USER, required=False),
+                Parameter("start", ParamType.DATE, required=True),
+                Parameter("end", ParamType.DATE, required=True),
+                Parameter("search_term", ParamType.STRING, required=False),
+                Parameter("fields", ParamType.LIST, required=False),
+            ),
+        )
+
+    def flatten_json(self, es_json: Dict[AnyStr, Any]) -> Dict[AnyStr, Any]:
+        """
+        Utility function for flattening the json.
+        We convert a nested data layer object into a new object with only one layer of key/value pairs.
+        Objects such as parent -> child are flattened as parent.child and parent -> array -> [obj1,objN]
+        are flattened as parent.array.0 = obj1 parent.array.N = objN
+        json_in =
+            {
+                    "@timestamp": "2021-07-12T14:08:46.563303",
+                    "run": {
+                        "controller": "dhcp31-171.perf.lab.eng.bos.redhat.com",
+                        "name": "pbench-user-benchmark_npalaska-dhcp31-171_2021.07.12T14.08.46"
+                    },
+                    "host_tools_info": [
+                        {
+                            "hostname": [{"test1":"dhcp31-171"}, {"test2":"dhcp31-195"}]
+                        }
+                    ]
+                }
+        json_out =
+            {
+                '@timestamp': '2021-07-12T14:08:46.563303',
+                'run.controller': 'dhcp31-171.perf.lab.eng.bos.redhat.com',
+                'run.name': 'pbench-user-benchmark_npalaska-dhcp31-171_2021.07.12T14.08.46',
+                'host_tools_info.hostname.test1': 'dhcp31-171',
+                'host_tools_info.hostname.test2': 'dhcp31-195'
+            }
+        """
+        if type(es_json) is not dict:
+            raise PostprocessError(
+                HTTPStatus.INTERNAL_SERVER_ERROR,
+                "Elasticsearch _source value is not a json",
+            )
+
+        def x(es_json: Any, prefix="", out={}):
+            if type(es_json) is dict:
+                for key in es_json:
+                    while prefix[-2:-1] == ".":
+                        prefix = prefix[:-1]
+                    x(es_json[key], prefix + key + ".", out)
+            elif type(es_json) is list:
+                i = 0
+                for element in es_json:
+                    if type(element) is dict:
+                        x(element, prefix + ".", out)
+                    else:
+                        x(element, prefix + str(i) + ".", out)
+                    i += 1
```

Comment:
I've discussed the inner `if`/`else` statement [elsewhere](https://github.com/distributed-system-analysis/pbench/pull/2381#discussion_r686201449), but I'm otherwise confused by your last comment, here.  If you use `enumerate()` to iterate over the elements in `es_json`, you get a tuple consisting of the index and the value, so you have them both without needing to do the counting yourself.

Of course, if we decide that we don't need to disambiguate duplicate output key values (e.g., because we create a list when there are collisions), then you are correct that we don't need to keep track of the list index, in which case we don't need to use `enumerate()` (or the inner `if`/`else` structure).

Reply:
This discussion continues [here](https://github.com/distributed-system-analysis/pbench/pull/2381#discussion_r688172586).